replit/codemirror-vim
# Block cursor improvements

# Why


The block cursor currently has a couple of issues.

Firstly, the cursor is taller on an empty line than on non-empty lines. To illustrate this, all of the following screenshots use `line-height: 2.5`.

Non-empty line:
![small_cursor](https://user-images.githubusercontent.com/158919/220490673-ecd3f484-75cb-408c-9a67-6dda5fec1af0.png)

Empty line:
![large_cursor](https://user-images.githubusercontent.com/158919/220490708-584d6f07-f92b-4576-aca5-c18e7c88495b.png)

Secondly, an unfocused cursor overlays a copy of the character such that both are visible, making the character look stronger:
![strong_m](https://user-images.githubusercontent.com/158919/220490914-9db0bef7-a4c0-4fa7-9ffc-8501c171baa9.png)



# What changed


For the first issue, I've added a `fullHeightCoordsAtPos` function, which acts like `coordsAtPos` except that the top and bottom coordinates represent the top and bottom of the whole visual line, not just the text. It also works with wrapped lines. However, in order to work with wrapped lines, it relies on the assumption that every visual line within the wrapped line has the same height, which is not necessarily the case if the line contains widgets or decorations that affect the height of the line.
![new_cursor](https://user-images.githubusercontent.com/158919/220491526-3af1555c-8891-4874-8af7-585961681cc4.png)

For the second issue, I've made the colour of the content of an unfocused block cursor transparent:
![normal_m](https://user-images.githubusercontent.com/158919/220491500-44dbf09a-9e78-4989-b43b-9118e1250381.png)



# Test plan



1. Add a line height of 2.5 to the editor in `dev/index.ts`: `EditorView.baseTheme({
    '.cm-line': {
      lineHeight: '2.5'
    }
  })`
2. View the dev page in a browser. Place the cursor on an empty line and a non-empty line to observe the difference.
3. Place the cursor over a letter. Click outside the editor. Look closely at the letter inside the cursor.
